include/libpmemobj++/experimental/persistent_concurrent_hash_map.hpp

```diff
@@ -0,0 +1,1716 @@
+                my_table[i] = pmemobj_oid( my_embedded_segment + segment_base( i ) );
+            my_mask = embedded_buckets - 1;
+            __TBB_ASSERT( embedded_block <= first_block, "The first block number must include embedded blocks" );
+        }
+
+        //! Initialize buckets
+        void init_buckets( segment_ptr_t ptr, size_type sz, bool is_initial ) {
+            persistent_ptr<node_base> rehashed_flag = is_initial ? rehashed : rehash_req;
+            bucket* b = ptr.get();
+            for ( size_type i = 0; i < sz; ++i, ++b ) {
+                __TBB_ASSERT(*reinterpret_cast<intptr_t*>(&b->mutex.get()) == 0, "Bucket mutex should be in unlocked state during initialization");
+                b->tmp_node = rehashed_flag;
+                b->node_list = empty_bucket;
+            }
+            get_pool_base().persist( ptr.get(), sizeof( ptr[0] )*sz );
+        }
+
+        void init_buckets(pool_base& pop, segment_facade_t& segment, bool is_initial) {
+            persistent_ptr<node_base> rehashed_flag = is_initial ? rehashed : rehash_req;
+            for (size_type i = 0; i < segment.size(); ++i) {
+                bucket* b = &(segment[i]);
+                __TBB_ASSERT(*reinterpret_cast<intptr_t*>(&b->mutex.get()) == 0, "Bucket mutex should be in unlocked state during initialization");
+                b->tmp_node = rehashed_flag;
+                b->node_list = empty_bucket;
+            }
+            // Flush in separate loop to avoid read-after-flush
+            for (size_type i = 0; i < segment.size(); ++i) {
+                bucket* b = &(segment[i]);
+                pop.flush(b, sizeof(bucket));
+            }
+            pop.drain();
+        }
+
+        //! Add node @arg n to bucket @arg b
+        void add_to_bucket( bucket *b, pool_base& pop ) {
+            __TBB_ASSERT( b->tmp_node != rehash_req, NULL );
+            __TBB_ASSERT( is_valid(b->tmp_node), NULL);
+            __TBB_ASSERT( b->tmp_node->next == b->node_list, NULL );
+            b->node_list = b->tmp_node; // its under lock and flag is set
+            pop.persist( &(b->node_list), sizeof( b->node_list ) );
+        }
+
+        //! Enable segment
+        void enable_segment( segment_index_t k, bool is_initial = false ) {
+            __TBB_ASSERT( k, "Zero segment must be embedded" );
+            pool_base pop = get_pool_base();
+            size_type sz;
+            if ( k >= first_block ) {
+#if INTEL_PRIVATE
+                TRACEP( "enable %u\n", unsigned( k ) );
+#endif
+                segment_facade_t new_segment(my_table, k);
+                sz = new_segment.size();
+                if (!new_segment.is_valid()) {
+                    new_segment.enable(pop);
+                }
+
+                init_buckets( pop, new_segment, is_initial );
+                sz <<= 1;// double it to get entire capacity of the container
+            }
+            else { // the first block
+#if INTEL_PRIVATE
```

Comment:
`#if DEBUG`
do we need this though? I think we should just remove these traces.